### DISCUSSION TITLE
-clahe producing bad results in webassembly (undefined behavior?)

### DISCUSSION THREAD

**Discussion by KnicKnic** [ImageMagick/ImageMagick#2454] [Help] [answered]
## Problem

When i run algorithm -clahe against images from an imagemagick build they produce corrupt results.

### Command

the command I used was `convert * -clahe 25x25%+128+3 -set filename:mysize %t %[filename:mysize]-%d.jpeg`

### Source / Good results
The image I used was the mountain image from https://imagemagick.org/script/clahe.php

### Bad output

the output that I got was [corrupted_mountains](https://i.imgur.com/mXHfBVq.jpg)
I used the following command / imagemagick to preform this action [webassembly action](https://knicknic.github.io/imagemagick/#!%7B"desc"%3A"-clahe%20shows%20corrupt%20image"%2C"cmd"%3A"convert%20*%20-clahe%2025x25%25%2B128%2B3%20-set%20filename%3Amysize%20%25t%20%25%5Bfilename%3Amysize%5D-%25d.jpeg"%2C"autoZip"%3Afalse%2C"magickEachPicture"%3Afalse%2C"showAdvanced"%3Afalse%7D)

### Theories

So webassembly does produce different results on certain unspecified behavior in C/C++. Are any tests for imagemagick (specifically clahe) with something like https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html for executing undefined behavior.

I ran with emscripten, however this caused too many local variables and imagemagick failed to load in wasm.

**Comment by fmw42** [comment]
Nick, just to inform you in case this information did not get to you. I tested the image that I sent to you originally (see below). It works on my Mac OSX Sierra (an old OS) using Chrome, Firefox and Opera, but failed with Safari. However, when I reduced the size to either 50% or 75% of the original dimensions, then it worked fine on all four browsers. So there is something that seems to me to be browser related. But I will defer to the ImageMagick developer for his input.

Have you checked on different browsers and different OS environments?

This is the file I tested.

![Test](https://user-images.githubusercontent.com/13458218/90950429-2f234a00-e406-11ea-91d3-45be1113975c.jpg)



Fred

**Comment by dlemstra** [comment, marked as answer]
Thanks for reporting this @KnicKnic. My [Magick.WASM](https://github.com/dlemstra/Magick.WASM) library does not contain this function yet but when I added this to the API and translated the command you provided I was able to reproduce the issue. But after changing the command to `convert logo: -clahe 640x480+128+3 output.jpg` it was much easier to track down why it failed because the first pixel had an incorrect value. The problem seems to be inside `InterpolateCLAHE` method in enhance.c:

```C
    for (x=(ssize_t) tile->width; x > 0; x--)
    {
      intensity=lut[*pixels];
      *pixels++=(unsigned short ) (PerceptibleReciprocal((double) tile->width*
        tile->height)*(y*(x*Q12[intensity]+(tile->width-x)*Q22[intensity])+
        (tile->height-y)*(x*Q11[intensity]+(tile->width-x)*Q21[intensity])));
    }
```

While debugging I changed the code to this:

```C
    for (x=(ssize_t) tile->width; x > 0; x--)
    {
      intensity=lut[*pixels];
      double a=PerceptibleReciprocal((double) tile->width*tile->height);
      double b=(x*Q12[intensity]+(tile->width-x)*Q22[intensity]);
      double c=(x*Q11[intensity]+(tile->width-x)*Q21[intensity]);
      *pixels++=(unsigned short ) (a*(y*b+(tile->height-y)*c));
    }
```

And suddenly the issue was gone. I don't understand why this solves the problem but I don't mind patching our code after these variables have some meaningfull names.

These where the values: 
```
tile->width: 320
tile->height: 240
intensity: 127
Q12[intensity]: 65131
Q22[intensity]: 65131
Q11[intensity]: 65131
Q21[intensity]: 65131
```

Invalid output: `9207.9466666666667152`
Correct output: `65131.000000000007276`

**Comment by dlemstra** [comment]
Turns out a typecast was missing. Just pushed a patch to resolve this.

**Reply by KnicKnic** [reply]
@dlemstra sounds like manual code debugging found this. Do you know if https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html would find this issue?

From the patch it looks like it indeed was undefined behavior that was the cause.

**Comment by urban-warrior** [comment]
We run various sanitizers against ImageMagick on a regular basis including undefined behavior.  We have not seen this particular issue raised in the past by any of the sanitizers.
